Fix for crash after joint connected node is set to null

RootKiller opened the pull request:
Hello!

In this PR I have fixed two issues with constraints/joints in Godot 4.

# Fix for crash after joint is destroyed
Commit: `Fixed crash in islands update after destroying`

Once I set up joint connecting two bodies and unset at least one of them from scripts
```gdscript
		joint.set_node_a(NodePath())
		joint.set_node_b(NodePath())
```
The engine was crashing because the dangling pointer to just removed the joint was still stored inside the constraint_map of the body. The crash was happening in `Step3DSW::_setup_island`

```cpp
void Step3DSW::_setup_island(Constraint3DSW *p_island, real_t p_delta) {
	Constraint3DSW *ci = p_island;
	while (ci) {
		ci->setup(p_delta);
		//todo remove from island if process fails
		ci = ci->get_island_next(); // <<<< get_island_next was pure virtual in my case
	}
}
```

I tested the fix on 100% repro. After the fix didn't observe any anomalies.

pouleyKetchoupp commented:
Please make two separate PRs instead of two commits since the changes are not directly related and fix two different issues. You can update this PR to handle one of them and create a separate one for the other change.

About the crash fix:
Could you share a minimal project to reproduce it? Removing constraints in the bodies is supposed to be handled in here for any type of joint:
https://github.com/godotengine/godot/blob/0e8fae1038cb27372777623b1af26e6f0cfb34fa/servers/physics_3d/physics_server_3d_sw.cpp#L1315-L1317
So I'm not sure why it's not done properly in your use case.

RootKiller commented:
Yes, the point is these joints are not destroyed. Here is zip with repro project. Once you start just press any key from 1 to 4 to detach the bodies. I'm not destroying the joints.

[ReproProject.zip](https://github.com/godotengine/godot/files/6054887/ReproProject.zip)

P.S.I will keep this PR as the crash fix and move the fix for collision to other one.

RootKiller commented:
> for (int i = 0; i < joint->get_body_count(); i++) { 
>  	joint->get_body_ptr()[i]->remove_constraint(joint); 
>  }

Perhaps if the way I suggested in PR is ok to fix it, we could remove these lines from `PhysicsServer3DSW::free`?

pouleyKetchoupp commented:
> Perhaps if the way I suggested in PR is ok to fix it, we could remove these lines from `PhysicsServer3DSW::free`?

Ah yes, actually I just found where the issue comes from. In 3.2, `free` is called each time a joint is re-configured in `Joint3D::_update_joint`, but not on master anymore.

So your fix makes sense and indeed the old code can be removed from `PhysicsServer3DSW::free`, although I would rather move this code to `Joint3DSW` destructor instead of having specific code in each joint type. It would allow possible mistakes in the future if new joint types are created, and avoid redundant code.

Also, if you don't mind, it would be nice to do the same fix in `Joint2DSW` as well since it has the exact same issue.

RootKiller commented:
"although I would rather move this code to Joint3DSW destructor instead of having specific code in each joint type."

Totally agree with that. Unfortunately, the pointers to the bodies are in child classes so I cannot remove them from there. What would you say is the way to go? I can, in theory, refactor it and move these members to Joint3DSW. I'm not sure if that is within the scope of this PR.

pouleyKetchoupp commented:
> 
> 
> "although I would rather move this code to Joint3DSW destructor instead of having specific code in each joint type."
> 
> Totally agree with that. Unfortunately, the pointers to the bodies are in child classes so I cannot remove them from there. What would you say is the way to go? I can, in theory, refactor it and move these members to Joint3DSW. I'm not sure if that is within the scope of this PR.

It should be possible to do it in a similar way `free` is doing it, unless I'm missing something:
```
for (int i = 0; i < get_body_count(); i++) {
    get_body_ptr()[i]->remove_constraint(joint);
}
```

RootKiller committed:
Move removal of the constraints from bodies to Joint3DSW destructor

Joint3DSW instances are also destroyed without explicitly calling free()
for example when changing constrainted bodies at runtime using
set_node method.

RootKiller commented:
Yes, good point. I'll do it this way.

RootKiller commented:
I've updated Joint3DSW.

I'm not sure about Joint2DSW though. It properly removes the constraint in the destructor so this part is fine.

And about refactoring it and making sure we remove it in the base class. I see that remove_constraint has explicit `p_pos` that has to match on both add_constraint/remove_constraint. The explicit removal here is IMHO increasing readability. I'm not sure if hiding it deeper is desired, unless we also hide the add_constraint so there is a clear logic of what the `p_pos` is and we don't give them freedom to chose the `p_pos` to the joint implementations.

RootKiller committed:
Refactor removal of constraints from bodies in 2D joints

RootKiller commented:
Ok, I'm reading it further and I guess p_pos MUST be equal to the position of the body in the body's array. So I can easily assume it will be correct.

pouleyKetchoupp commented on servers/physics_2d/joints_2d_sw.h:
```diff
@@ -60,6 +60,12 @@ class Joint2DSW : public Constraint2DSW {
+
+	virtual ~Joint2DSW() {
+		for (int i = 0; i < get_body_count(); i++) {
+			get_body_ptr()[i]->remove_constraint(this, i);
```
It should probably check for the pointer to be valid for each body, since that's what `PinJoint2DSW` was doing. In general, it would be safer to do that for both 2D and 3D joints because in some cases one of the bodies can be null (I think this is currently supported only with 3D joints, but it might change in the future).

RootKiller commented:
Review addressed

pouleyKetchoupp reviewed (approved):
Looks good!

akien-mga commented:
Thanks!
